src/extent.c

```diff
@@ -2162,24 +2164,58 @@ extent_merge_default_impl(void *addr_a, void *addr_b) {
 	return false;
 }
 
-#ifdef JEMALLOC_MAPS_COALESCE
+static bool
+extent_head_no_merge(extent_t *a, extent_t *b) {
+	/*
+	 * When coalesce is not always allowed (Windows), only merge extents
+	 * from the same VirtualAlloc region under opt.retain (in which case
+	 * MEM_DECOMMIT is utilized for purging).
+	 */
+	if (maps_coalesce) {
+		return false;
+	}
+	if (!opt_retain) {
+		return true;
+	}
+	/* If b is a head extent, disallow the cross-region merge. */
+	if (extent_is_head_get(b)) {
+		/*
+		 * Additionally, sn should not overflow with retain; sanity
+		 * check that different regions have unique sn.
+		 */
+		assert(extent_sn_comp(a, b) != 0);
+		return true;
+	}
+	assert(extent_sn_comp(a, b) == 0);
+
+	return false;
+}
+
 static bool
 extent_merge_default(extent_hooks_t *extent_hooks, void *addr_a, size_t size_a,
     void *addr_b, size_t size_b, bool committed, unsigned arena_ind) {
+	if (!maps_coalesce) {
```

Comment:
I think this logic should live in extent_merge_default_impl, no? It feels weird that we're splitting up the coalescing logic this way.

Reply:
I have been back and forth on this a couple of times; the reason I wanted `extent_head_no_merge` to be outside of `extent_merge_default_impl` is to make sure in `extent_merge_impl`, we check the extents mergeability before calling the merge hook, so that if user has custom merge hook, they won't get two unmergeable extents (since the head state is tracked internally). I do agree the split logic this way is strange.

How about, we move the `extent_head_no_merge` into `extent_merge_default_impl`; but in the custom merge hook branch here https://github.com/jemalloc/jemalloc/pull/1545/files#diff-7b26b977303fe92c093a2245b0eaf255R2239, we check `extent_head_no_merge` before calling the user hook?

Reply:
I sort of thought we wouldn't track head state at all for extent-hook-managed extents; it seems like keeping track of that is the user's job.

I don't think that there's a good way of separating this out though, especially in the presence of potentially-mutable extent hooks. So I think you're right that the current way is better.

We should probably add some mechanism for this to the extent hook wishlist, though.

Reply:
Okay I won't change this part then. Update coming soon; will add it to the wishlist as well.

Reply:
Just updated @davidtgoldblatt